### ISSUE TITLE
"Look-Ahead" Taint Analysis

### ISSUE THREAD

**Issue by luigidcsoares** [lac-dcc/lif#9]
For side-channel _verification_, the taint analysis considers both data and control dependencies. However, by linearizing secret-dependent branches we remove control dependencies, so our taint analysis should consider only control dependencies that remain in the program after linearization (but we need the taint analysis in order to be able to linearize something, hence the "look-ahead" term).

Consider the following piece of code:

```C
if (use(secret)) { 
  // do something 
} else {
  // do something else
}
// pdom
```

The predicate that controls the if statement is data dependent on a secret. Thus, that if statement will be linearized and any control dependency from the predicate to instructions inside the _then_ or the _else_ paths will be removed as long as the effects of such instructions don't _escape the influence region_ of said predicate (i.e., they don't affect anything from pdom onwards). The question then is: when does something escape the influence region of a predicate?

To the best of my knowledge, there are only two scenarios to consider:

a) phi functions

```C
if (use(secret)) {
  x0 = 0
} else {
  x1 = 1
}

x = phi x0, x1
if (use(x)) {
  // do something
}
```

We already handle phi functions!

b) store + load

```C
if (use(secret)) {
  *x = 0
} else {
  *x = 1
}

if (use(*x)) {
  // do something
}
```

This is currently not implemented. There are benchmarks in our test suite that contain this pattern. One example is [ghostrider/dijkstra](https://github.com/lac-dcc/lif/blob/b5680653eceaf67d6d2090f87d663c6727fc5525/lif/test/benchmarks/scelim-2018/ghostrider/dijkstra/lib/dijkstra.c#L4). To circumvent this problem, we currently identify and mark variables as secret by hand. In the case of `dijkstra`, we had to mark the vectors `vis` and `dis` as secret. This process is done as follows:

```C
#define secret __attribute__((annotate("secret")))
secret int secret_var;
```

**Comment by pronesto** [comment]
Hi Luigi, if you have something like:


```
j0 = 0;
HEAD:
  j = phi(j0, j1)
  p = ??
  if p goto OUT
  ...
  j1 = ??
  goto HEAD
OUT:
  ?? = j
```

Is it not the case that `j` escapes the influence region of `p` without a phi-function outside this region? Or do you consider the phi-function inside it already to be enough? I think that within the loop, `j` is not tainted, but outside it, it is tainted.

**Comment by luigidcsoares** [comment]
> Is it not the case that j escapes the influence region of p without a phi-function outside this region? 

Yes, but we currently work with loops in LCSSA form (https://llvm.org/docs/LoopTerminology.html#loop-closed-ssa-lcssa). For that, we first apply the `-lcssa` pass, which splits the live range of variables that outlive loops.

In your example, a phi function would be inserted in the basic block OUT:

```C
j0 = 0;
HEAD:
  j = phi(j0, j1)
  p = ??
  if p goto OUT
  ...
  j1 = ??
  goto HEAD
OUT:
  j.out = phi(j)
```

and `j` is replaced with `j.out` outside the loop. Now there is a control dependency from `p` to the definition of `j.out`.
